src/libcore/slice/mod.rs

```diff
@@ -1575,6 +1575,127 @@ impl<T> [T] {
         sort::quicksort(self, |a, b| f(a).lt(&f(b)));
     }
 
+    /// Rearranges the elements of the slice such that the following properties hold:
+    ///
+    /// - The value at position `index` will be the same as if the array were sorted.
+    /// - Any value at position `i < index` will be less than or equal to any value at a position
+    ///   `j > index`.
+    ///
+    /// This reordering is unstable (i.e. any number of equal elements may end up at position
+    /// `index`), in-place (i.e. does not allocate), and `O(n)` worst-case.
+    ///
+    /// # Current implementation
+    ///
+    /// The current algorithm is based on the quickselect portion of the same quicksort algorithm
```

Comment:
I would mention that this can be used to find the median of a set with `slice.sort_by_index(slice.len() / 2)`,
also mention that this might be called "kth element" in other libraries.

Reply:
I've updated the documentation. PTAL.

Reply:
I was thinking an entry in the examples section, unless that's overkill. Something like:

```rust
fn find_median(values: &mut [i32]) -> f32 {
    // median of an empty vector is undefined
    if (values.len() == 0) { return ::std::f32::NAN }

    let half_len = values.len() / 2;

    // if the slice is even-length, the median is the average of the two middle values
    if (values.len() % 2 == 0) {
        values.sort_at_index(half_len - 1);
        // find the second middle value, leaving the first one inplace
        values[half_len..].sort_at_index(0);
        (values[half_len - 1] + values[half_len]) as f32 / 2.0
    ) else {
        values.sort_at_index(half_len);
        values[half_len] as f32
    }
}

fn main() {
    assert_eq!(find_median(&mut [1, 5, 2, 4, 3], 3.0);
    assert_eq!(find_median(&mut [6, 5, 4, 3, 2, 1], 3.5);
}
```

Reply:
Did you mean to add this to the examples section of the documentation of each version of the function? This seems somewhat larger than what we have in most other examples (and as written tends to use a lot more rust features than simply calling `sort_at_index`). Is there a different place to put this example? Perhaps in the `doc/rust-by-example` folder?

In the examples that *are* provided in the functions' documentation, we're already taking the median of a five-element list. Would it be better to add a comment in that example clarifying that it is the median?

Reply:
You're right, it's probably better to put the extended example somewhere else.

>In the examples that are provided in the functions' documentation, we're already taking the median of a five-element list. Would it be better to add a comment in that example clarifying that it is the median?

I would say so, yeah. Also clarify why you're matching five different array patterns (i.e. showing that it only guarantees the placement of the selected element).

Reply:
I've added comments to the examples. Note, it also guarantees the elements to the left and right of the median, but not their order.